### PR TITLE
Developer Documentation

### PR DESCRIPTION
**Pull request by bilbo7833**
## Documentation
Renamed project to Acorn
Added READMEs on how to get started with AcornAgent.
Also added examples at different levels of abstraction.

### CODE REVIEW

File: java/apirag-starter/pom.xml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      <artifactId>slf4j-simple</artifactId>
+      <version>2.0.13</version>
+    </dependency>
+
+    <!-- Testing -->
+    <dependency>
+      <groupId>com.datasqrl</groupId>
+      <artifactId>apirag-openai</artifactId>
+      <version>${project.version}</version>
+<!--      <scope>test</scope>-->
```
Comment (bilbo7833) [inline java/apirag-starter/pom.xml:53]: @mbroecheler Whatever model provider we use in the examples needs to be a full dependency here

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  /**
+   * Initializes a command line chat bot using OpenAI
+   *
+   * @param toolsPath     A path to a tools file in JSON format
+   * @param apiExecutor   An APIExecutor for the tools
+   * @param systemPrompt  The system message for the LLM
+   * @param openAIKey     The OpenAI API key to call the API
+   */
+
+  public LowLevelAgent(Path toolsPath, APIExecutor apiExecutor, String systemPrompt, String openAIKey) throws IOException {
```
Comment (bilbo7833) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:60]: @mbroecheler Pls review this low level agent code carefully. I am not sure that I got the abstraction level right. Should I use the backend directly to validate the function call? Should I introduce the retry logic here? Or anything else that should be at a different abstraction level?

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+              }
+              if (accumulator.getMessageChunk().getContent() != null) {
+                System.out.print(accumulator.getMessageChunk().getContent());
+              }
+            }
+          })
+          .doOnComplete(System.out::println)
+          .lastElement()
+          .blockingGet()
+          .getAccumulatedMessage();
```
Comment (bilbo7833) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:115]: @mbroecheler This part of the code makes it complicated. I don't even understand what is happening here. Do you think we should keep this?

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java (hunk 2 above)
Comment (mbroecheler) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:60]: I like that this uses the streaming API (even though that makes the code more complex it highlights that you have that low level control).
My suggestions:
1. Create a class MyChatSession that extends ChatSession to overwrite the `getContextWindow(int maxTokens, ModelAnalyzer<Message> analyzer)` function with custom context window creation logic. I think that's a piece a lot of people want fine grained control over (since it determines the prompt and hence the cost & quality). Do a simple implementation that only ever takes up to the last 2 messages into the context.
2. Add a comment in there saying that you'd likely want to add retry logic for failed function calls and point to ChatProvider implementation that does this (no need to copy the code here).

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java (hunk 3 above)
Comment (mbroecheler) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:115]: Yes, I like it ;-)

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    String systemPrompt = "You are an incredibly enthusiastic and joyous life coach. You provide advice to people. You look up any and all advice you give via the provided functions.";
+    Path toolsPath = Path.of("java", "apirag-starter", "src", "main", "resources", "tools", "advice.tools.json");
+    String openaiKey = System.getenv("OPENAI_API_KEY");
+    APIExecutor apiExecutor = new RESTExecutorFactory().create(new MapConfiguration(Map.of(
+        "type", "rest",
+        "url", "https://api.adviceslip.com"
+    )), "adviceapi");
+
+    LowLevelAgent agent = new LowLevelAgent(toolsPath, apiExecutor, systemPrompt, openaiKey);
+    Map<String, Object> context = Map.of("userid", 1);
```
Comment (mbroecheler) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:152]: Remove the context from this example. It's not needed. Pass a Map.of() to the super class in the constructor of MyChatSession.

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        ChatMessage nextMsg = new UserMessage(nextLine);
+        session.addMessage(nextMsg);
+      }
+    }
+  }
+
+  public static void main(String... args) throws Exception {
+    String systemPrompt = "You are an incredibly enthusiastic and joyous life coach. You provide advice to people. You look up any and all advice you give via the provided functions.";
+    Path toolsPath = Path.of("java", "apirag-starter", "src", "main", "resources", "tools", "advice.tools.json");
```
Comment (mbroecheler) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:144]: This path looks strange to me. If you run this in Intellij it should be just relative to the module path, no?

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+ * A simple streaming chatbot for the command line.
+ */
+@Slf4j
+@Value
+public class LowLevelAgent {
+
+  String systemPrompt;
+  OpenAiService service;
+  ToolsBackend toolsBackend;
+  OpenAIModelConfiguration chatConfig = new OpenAIModelConfiguration(
```
Comment (mbroecheler) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:46]: IMHO we should do all configuration in the main().

File: java/apirag-starter/src/main/java/com/datasqrl/ai/ConfigurableChatBot.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      if (nextLine.equalsIgnoreCase("exit")) {
+        System.exit(0);
+      }
+      message = nextLine;
+    }
+  }
+
+  public static void main(String... args) throws Exception {
+    String systemPrompt = "You are a huge fan of the Ricky and Morty TV show and help users answer questions about the show. You always try to look up the information a user is asking for via one of the available functions. Only if you cannot find the information do you use general knowledge to answer it. Retrieved information always takes precedence. You answer in the voice of Jerry Smith.";
+    List<RuntimeFunctionDefinition> tools = ToolsBackendFactory.readTools(Path.of("java", "apirag-starter", "src", "main", "resources", "tools", "rickandmorty.tools.json"));
```
Comment (mbroecheler) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/ConfigurableChatBot.java:65]: This path doesn't look right to me.

File: java/apirag-starter/src/main/java/com/datasqrl/ai/ConfigurableChatBot.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    )), "rickandmortyapi");
+    ToolsBackend toolsBackend = ToolsBackendFactory.of(tools, Map.of(APIExecutorFactory.DEFAULT_NAME, apiExecutor));
+    ChatProvider<?, ?> chatProvider = ChatProviderFactory.fromConfiguration(Map.of(
+        "provider", "openai",
+        "name", "gpt-3.5-turbo",
+        "temperature", 0.8
+    ), toolsBackend, systemPrompt);
+
+    ConfigurableChatBot chatBot = new ConfigurableChatBot(chatProvider);
+    Map<String, Object> context = Map.of("userid", 1);
```
Comment (mbroecheler) [inline java/apirag-starter/src/main/java/com/datasqrl/ai/ConfigurableChatBot.java:78]: We don't need the context here.

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java (hunk 4 above)
Comment (bilbo7833) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:152]: ok, but this is one of the things that is hardest to understand. your implementation of ContextKeys in the AcornAgentServer is also not very easy to understand why the context is there for, since it's not often met in ChatBot examples

File: java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java (hunk 5 above)
Comment (bilbo7833) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:144]: I tried, but if I run the java file by itself, it starts in the home folder. 🤷

Comment (mbroecheler) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:144]: Got it. How are you running it?

Comment (bilbo7833) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java:144]: Screenshot 2024-07-17 at 22 05 39
LIke this

**Comment by bilbo7833** [comment]
@mbroecheler 

1. Need to update two diagrams for the front page ✅ 
2. Rebase on main (to add vertex) and add README.md with documentation ✅ 
3. Rename all the modules from “apiRAG” to “acorn” (and elsewhere if there are any mentions of “apiRAG” left) ✅ 
4. Update docker image (to acorn:0.5) and release to sonatype

File: java/apirag-starter/src/main/java/com/datasqrl/ai/ConfigurableChatBot.java (hunk 7 above)
Comment (bilbo7833) [reply java/apirag-starter/src/main/java/com/datasqrl/ai/ConfigurableChatBot.java]: @mbroecheler How would this path look like if you were to import just the Maven starter dependencies or if you were to run the Java file locally?

File: README.md [hunk 9]
```diff
@@ -1,66 +1,117 @@
...
-It extends OpenAI's function call configuration with additional information on how to execute the function against an existing API.
+## Acorn Agent Features
 
-For more information on apiRAG's configuration format, check out the [format documentation](FORMAT.md).
+* On-Demand Data Retrieval from GraphQL APIs, REST APIs, or JDBC-compatible databases.
+* Pluggable models and model providers like OpenAI (GPT models), Bedrock (Llama, and others), Google Vertex (Gemini), and Groq (Llama, Mixtral, and others).
+* Message history to preserve context between agent interactions.
+* Custom data visualizations.
+* Easy integration with Spring Boot and other web development frameworks with in-code or file-based configuration.
+* UI integration through function callbacks.
```
Comment (bilbo7833) [inline README.md:39]: @mbroecheler What do you mean by this point? Function Callbacks is a bit confusing here when we talk about a function calling platform.

File: examples/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+You can run the following examples in Docker by standing up an API and running Acorn Agent as a server in front of it:
+
+* [Customer Credit Card](https://github.com/DataSQRL/datasqrl-examples/tree/main/finance-credit-card-chatbot): Builds a ChatBot that analyzes a customer's credit card transactions to answer questions about spending.
+* [IoT Bot](https://github.com/DataSQRL/datasqrl-examples/tree/main/iot-sensor-metrics): A data agent for retrieving and analyzing metrics in realtime.
+* [Credit Card Rewards](https://github.com/DataSQRL/datasqrl-examples/tree/main/finance-credit-card-chatbot): An agent that helps customers pick the best credit card to maximize rewards.
+
+For these examples, you will stand up an API locally that Acorn Agents instruments for LLM usage. Hence, these examples are a little more involved to run than the public API examples below. However, because we have control over the API, we can use more of Acorn Agent's feature set like storing and retrieving user messages, adding more complex tools for data retrieval, and custom data visualization.
+
+These examples show you what building a production-grade application with Acorn Agent looks like.
```
Comment (bilbo7833) [inline examples/README.md:15]: @mbroecheler But these examples no longer live in this repo. Do you plan to include a link to the datasqrl-examples repo or to move them back in here?

File: java/acorn-spring/README.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+Compile the acorn-spring package
+```
+mvn package
+```
+
+Run the `AcornAgentServer` with the configuration parameters and the API Keys
+```
+java -jar /app/server.jar --agent.config=[CONFIG_FILE] --agent.tools=[TOOLS_FILE] -DOPENAI_API_KEY=XXXX
+```
```
Comment (bilbo7833) [inline java/acorn-spring/README.md:19]: @mbroecheler How do you run a spring boot application locally? I couldn't get this to run on my command line yet

Comment (mbroecheler) [reply java/acorn-spring/README.md:19]: What error do you get?

File: README.md (hunk 9 above)
Comment (mbroecheler) [reply README.md:39]: clarified.

File: examples/README.md (hunk 10 above)
Comment (mbroecheler) [reply examples/README.md:15]: I added a link.

File: TOOLS_CONFIG.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+* `api`: Defines the API query that executes the function and returns the requested data. The `api` config object has the following fields:
+  * `name`: Name of the API which must match a configured APIExecutor. Defaults to `default`.
+  * `query`: Query to execute (for GraphQL and JDBC) with parameters that match the function parameters.
+  * `path` and `method`: The REST method (GET, POST, etc) and path for the REST resource. The path can contain path arguments that match the function arguments.
+
+## Special Functions
+
+Acorn Agent defines the following special functions which are used internally for message persistence and not passed to the LLM:
+
+* `InternalSaveChatMessage`: The API query to store a chat message. This is used to persist the chat history in the API for future retrieval and/or analysis. A chat message has the following fields:
```
[the remark this answers is not in the document]
Comment (mbroecheler) [reply TOOLS_CONFIG.md:16]: Great point. I added explicit examples and pointed to the test cases that contains an example.

File: java/acorn-graphql/src/test/resources/graphql/nutshop-schema.tools.json [hunk 13]
```diff
@@ -12,7 +12,7 @@
         },
         "customerid" : {
           "type" : "integer",
-          "description" : " The unique identifier for the customer."
+          "description" : "customerid: The unique identifier for the customer."
```
Comment (bilbo7833) [inline java/acorn-graphql/src/test/resources/graphql/nutshop-schema.tools.json:15]: Is this correct?

**Review by bilbo7833 — COMMENTED** [review]
Looks good, apart from one comment

File: java/acorn-graphql/src/test/resources/graphql/nutshop-schema.tools.json (hunk 13 above)
Comment (mbroecheler) [reply java/acorn-graphql/src/test/resources/graphql/nutshop-schema.tools.json:15]: Yes, this is an edge case that triggered a bug: if the user includes the field name in the comment it caused an issue. This is fixed now and I added it here as a test case.

**Review by mbroecheler — APPROVED** [review]
